src/business-logic/services/timeout-callback-scheduler.ts

```diff
@@ -1,27 +1,29 @@
 import { CallbackScheduler } from './interfaces/callback-scheduler'
+import { ConsoleLogger } from '../../console-logger'
+import { Logger } from '../../logger'
 
 const EXECUTE_CALLBACK_DURATION_THRESHOLD_IN_MS: number = 1
 const SCHEDULE_RESOLUTION: number = 2
 
 export class TimeoutCallbackScheduler implements CallbackScheduler {
   private static instance: CallbackScheduler
 
-  public static getInstance(): CallbackScheduler {
+  public static getInstance(logger: ConsoleLogger): CallbackScheduler {
     if (!this.instance) {
-      this.instance = new TimeoutCallbackScheduler()
+      this.instance = new TimeoutCallbackScheduler(logger)
     }
     return this.instance
   }
-
+  private readonly logger: Logger
   private timeoutIdentifier?: NodeJS.Timeout
 
-  private constructor() {
-    return
+  private constructor(logger: ConsoleLogger) {
+    this.logger = logger.tag(TimeoutCallbackScheduler.name)
```

Comment:
Instead of setting the intance variable you can just do:
```
private constructor(private readonly logger: Logger) {
    logger.tag(TimeoutCallbackScheduler.name)
}
```

Note: You have made this class dependent on the concrete implementation `ConsoleLogger` instead of the abstraction: `Logger`. That's a no go.

Reply:
To elaborate on the invalidity. `logger.tag` creates a new logging context, so the suggestion creates a new logging context with the tag associated, and marks it as garbage instantly.